proposals/text/0000-shaper-in-prompt-template.md

```diff
@@ -0,0 +1,153 @@
+    ```
+
+This would set the default delimiter `\n` as defined in the template to `MY_NEW_DELIMITER` for this specific PromptNode.
+
+# Motivation
+
+Currently using PromptNode is a bit cumbersome as:
+- for using it in popular use-cases like question-answering, it requires to add the Shapers to the pipeline manually which creates a lot of boilerplate code and is not very intuitive
+- to customize a prompt within a pipeline, you may need to change four different things: the prompt node, the prompt template, the input shapers and the output shapers. This is not ideal as it requires to write a lot of boilerplate code and makes it hard to iterate quickly on prompts.
+- if you wanted to share your prompt template with the community, you would need to share the whole pipeline (as you do need shapers), which is not ideal as it may contain other nodes that are not relevant to the prompt template.
+
+
+# Detailed design
+
+PromptTemplate gets two new attributes: `input_shapers` and `output_shapers`. These are lists of Shaper objects that are applied to the input and output of the prompt respectively. The `input_shapers` are applied to the input of the prompt, the `output_shapers` are applied to the output of the prompt:
+PromptNode calls `PromptTemplate.prepare` before exectuting the prompt. `PromptTemplate.prepare` makes all `input_shapers` run on the `invocation_context`.
+PromptNode invokes the prompt on the prepared `invocation_context`.
+PromptNode calls `PromptTemplate.post_process` after executing the prompt. `PromptTemplate.post_process` makes all `output_shapers` run on the `invocation_context`.
+
+Note, that `Shapers` are still usable in Pipelines as before.
+
+# Drawbacks
+
+Look at the feature from the other side: what are the reasons why we should _not_ work on it? Consider the following:
+
+- What's the implementation cost, both in terms of code size and complexity? A good day
+- Can the solution you're proposing be implemented as a separate package, outside of Haystack? No
+- Does it teach people more about Haystack? No, but it makes it easier to use especially for beginners.
+- How does this feature integrate with other existing and planned features? It doesn't change any existing features and should nicely integrate with agents.
+- What's the cost of migrating existing Haystack pipelines (is it a breaking change?)? None
+
+It also fosters a bit the nesting of components in Haystack. Although the whole PromptNode ecosystem already does this (e.g. via PromptModel, PromptTemplate being used by PromptNode), it's still a bit of a new concept. However, I think it's a good one and it's not too hard to understand.
+
+We still don't have access to PromptNode, PromptModel or the invocation layer inside of PromptTemplates. If we want Shapers to access fundamental parts of them (e.g. the tokenizer), we would need to pass them to the Shapers. This would make the whole system more complex, but it would be possible.
```

Comment:
I'm wondering if this consideration would be relevant for truncating prompts in an "intelligent" way. By this I mean it would be nice if we could support truncation of prompts by shortening the number of inputs (e.g. `documents`) given to the `PromptTemplate.fill`. Currently we are adding support for "naive" truncation by cutting off the text at the end of the prompt that goes past the limit. See this [PR](https://github.com/deepset-ai/haystack/pull/4179).

I think this is would be difficult to do outside of `PromptTemplate` because it is hard to know which input variables in a user defined `PromptTemplate` should be truncated.

Reply:
Appealing, but let's leave it for the ultimate fine-tuning once everything is mature?

Reply:
I would say without this feature this would make me preferentially use the `OpenAIAnswerGenerator` for retrieval augmented question answering because then I can be confident that instructions at the end of my prompt aren't removed. And as @TuanaCelik experienced having instructions at the end of the prompt removed can cause really unexpected behavior in the model output.

Unfortunately, just setting a low top_k (e.g. 3) doesn't always guarantee this won't happen, which is why as a user I'd prefer the answer generator node at the moment.

Reply:
Ok cool, I see, I better understand the motivation now!

Reply:
Indeed this would be one of the advantages to have it in PromptTemplate. And it shouldn't be super hard to implement in this design